certora/specs/ERC20.spec

```diff
@@ -0,0 +1,252 @@
+│ Invariant: balance of address(0) is 0                                                                               │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+invariant zeroAddressNoBalance()
+    balanceOf(0) == 0;
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rules: only the token holder (or a permit) can increase allowance. The spender can decrease it by using it          │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule onlyHolderOfSpenderCanChangeAllowance(env e, method f){
+    requireInvariant totalSupplyIsSumOfBalances();
+
+    calldataarg args;
+    address holder;
+    address spender;
+
+    uint256 allowanceBefore = allowance(holder, spender);
+    f(e, args);
+    uint256 allowanceAfter = allowance(holder, spender);
+
+    assert (
+        allowanceAfter > allowanceBefore
+    ) => (
+        (f.selector == sig:approve(address,uint256).selector           && e.msg.sender == holder) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+
+    assert (
+        allowanceAfter < allowanceBefore
+    ) => (
+        (f.selector == sig:transferFrom(address,address,uint256).selector && e.msg.sender == spender) ||
+        (f.selector == sig:approve(address,uint256).selector              && e.msg.sender == holder ) ||
+        (f.selector == sig:permit(address,address,uint256,uint256,uint8,bytes32,bytes32).selector)
+    );
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: transfer behavior and side effects                                                                            │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule transfer(env e) {
+    requireInvariant totalSupplyIsSumOfBalances();
+    require nonpayable(e);
+
+    address holder = e.msg.sender;
+    address recipient;
+    address other;
+    uint256 amount;
+
+    // cache state
+    uint256 holderBalanceBefore    = balanceOf(holder);
+    uint256 holderVotingPowerBefore = delegatedVotingPower(delegatee(holder));
+    uint256 recipientBalanceBefore = balanceOf(recipient);
+    uint256 recipientVotingPowerBefore = delegatedVotingPower(delegatee(recipient));
+    uint256 otherBalanceBefore     = balanceOf(other);
+
+    // run transaction
+    transfer@withrevert(e, recipient, amount);
+
+    // check outcome
+    if (lastReverted) {
+        assert holder == 0 || recipient == 0 || amount > holderBalanceBefore ||
+            // Handle overflows in delegation, should not be possible.
+            recipientVotingPowerBefore + amount > max_uint256 || holderVotingPowerBefore < amount ;
```

Comment:
This can be avoided if we merge `ERC20.spec` and `Delegation.spec`, it's still unclear to me why though. I have this version in a stashed diff so I can push a commit with this version if needed. As well, the invariants, that are imported in `Delegation.spec`, could be moved to `Delegation.spec` and import `Delegation.spec` in `ERC20`. The initial goal was to make the minimal amount of changes to `ERC20.spec` with respect to the original version taken from the OZ repo. It feels weird to import the `Delegation.spec` in the `ERC20.spec` though. Suggestions to organize the code in an alternate way are welcome.

Reply:
I think I understood why merging both files doesn't revert. This is because the `_moveDelegateVotes` is summarized in `Delegation.spec` and then can't revert, not easy to spot.